firedrake/locate.c

```diff
@@ -25,9 +32,33 @@ int locate_cell(struct Function *f,
         if (f->extruded == 0) {
             for (uint64_t i = 0; i < nids; i++) {
                 if ((*try_candidate)(data_, f, ids[i], x)) {
+                    /* Found cell! */
                     cell = ids[i];
                     break;
                 }
+                else {
+                    /* Cell not found, but could be on cell boundary. We therefore look
+                       at our reference coordinates and find the point closest to being
+                       inside the reference cell. If we don't find a cell using try_candidate
+                       we assume that this process has found our cell. */
+                    /* First we must find the the reference coordinate which is closest to
+                       the interval [0, 1]. Note that ref_coords was updated as data_ by
+                       try_candidate. */
+                    for (uint64_t j = 0; j < dim; j++) {
+                        if(ref_coords->X[j] < 0.0){
+                            current_closest_ref_coord = ref_coords->X[j];
```

Comment:
Logic here is wrong if there is more than one negative coordinate (as happens if we're outside the corner of the cell). Should be `< current_closest_ref_coord` on line 48 and 52, and current_closest_ref_coord should be reset to 0.0 before line 47.